### ISSUE TITLE
"memoryCacheSize must be a positive number"

### ISSUE THREAD

**Issue by alewang** [Appboy/appboy-android-sdk#66]
We have been running some Android tests using Robolectric, and we are seeing the error presented in the title.

I believe this is caused in the com.appboy.Appboy class. During initialization of the Appboy object, it determines the cache size it is allowed to use by retrieving the maximum memory the JVM is allowed to use through Runtime.getRuntime().maxMemory() call. The returned value is then immediately cast to a primitive int type. maxMemory() returns a primitive long value. The evaluation of the cast from long to int can generate a negative integer, which leads to an error that is the title of this message.

**Comment by Bucimis** [comment]
Hi @alewang,

Thanks for reporting.  What version of the sdk are you running and what does Runtime.getRuntime().maxMemory() return on your test machine?  That will help us nail down exactly what happened since the image cache code has gone through a few revisions.

Depending on our findings, we will incorporate a fix into a release in the near future.

For now, it's actually possible to disable UIL altogether by setting `com_appboy_disable_uil_image_cache` to `true` in your `appboy.xml`.  Unless you are explicitly testing the Appboy newsfeed, this should solve the issue.

Thanks,
Appboy Android Team

**Comment by Bucimis** [comment]
Hi @alewang,

Just wanted to check in - were you able to successfully disable the UIL cache for your tests?

**Comment by Bucimis** [comment]
Hi @alewang,

I'm going to close this issue for now.  Please let us know if the issue is not resolved on your end or you have any further questions or issues.

We'll inform you of any updates to our UIL handling.

Thanks!

**Comment by ChristianToMidiOut** [comment]
@Bucimis I am having the same issue. Here is the stacktrace.

java.lang.IllegalArgumentException: memoryCacheSize must be a positive number

	at bo.app.id.b(SourceFile:359)
	at com.appboy.Appboy.<init>(SourceFile:195)
	at com.appboy.Appboy.getInstance(SourceFile:153)
	at com.playphone.appboywrapperlibrary.AppboyWrapper.changeUser(AppboyWrapper.java:87)
	at com.playphone.airtellite.tracking.AppboyTrackingService.initialize(AppboyTrackingService.java:34)
	at com.playphone.airtellite.tracking.Tracker.initTrackingApis(Tracker.java:49)
	at com.playphone.airtellite.activities.BaseActivity.onCreate(BaseActivity.java:78)
	at com.playphone.airtellite.activities.PresenterBaseActivity.onCreate(PresenterBaseActivity.java:32)
	at android.app.Activity.performCreate(Activity.java:5990)
	at org.robolectric.util.ReflectionHelpers.callInstanceMethod(ReflectionHelpers.java:195)
	at org.robolectric.util.ActivityController$1.run(ActivityController.java:126)
	at org.robolectric.shadows.ShadowLooper.runPaused(ShadowLooper.java:356)
	at org.robolectric.shadows.CoreShadowsAdapter$2.runPaused(CoreShadowsAdapter.java:40)
	at org.robolectric.util.ActivityController.create(ActivityController.java:123)
	at org.robolectric.util.ActivityController.create(ActivityController.java:133)
	at org.robolectric.util.ActivityController.setup(ActivityController.java:227)
	at org.robolectric.Robolectric.setupActivity(Robolectric.java:102)

**Comment by Bucimis** [comment]
Hi @ChristianToMidiOut,

Thanks for reporting this to us.

Since it appears you're using Robolectric, you should be able to apply a similar fix to the one recommended above.  For your test builds, please set `com_appboy_disable_uil_image_cache` to `true` in your `appboy.xml` and let us know if that solves the issue.

We are also addressing the root cause of this issue and will have a fix out in our next release.

Thanks,
Jared

**Comment by ChristianToMidiOut** [comment]
@Bucimis Im still getting the issue ever after setting that value in the the appboy.xml.

**Comment by Bucimis** [comment]
Hi @ChristianToMidiOut,

That's surprising to hear. Can you sanity check the build by cleaning it and making sure the new appboy.xml is used/read by your application?

There is also an alternative way of turning off UIL introduced in Appboy SDK 1.17.0.  In your `Application.onCreate()`, before any other calls to Appboy, you may call:
```
    AppboyConfig appboyConfig = new AppboyConfig.Builder()
        .setDisableUilImageCache(true)
        .build();
    Appboy.configure(this, appboyConfig);
```

Let us know if these steps help fix the issue.

Thanks,
Jared

**Comment by alewang** [comment]
Hi @Bucimis 

I was able to reproduce this error after setting the maximum heap size for the testing JVM to 3 GiB by adding -Xmx3072m to the unit test options in Android Studio.

Looking through the Appboy jar shows this line in the Appboy constructor:

int var5 = Math.min((int)Runtime.getRuntime().maxMemory() / 32, 1048576);

The cast of the return value from maxMemory() (returns a long) to an int causes the value to overflow, so we're passing a negative value in to Math.min(...).

We're running Appboy library version 1.13.5. It seems like the issue was resolved by version 1.17.0. I can not change the library version at the immediate moment, so I will keep a mental note of limiting the JVM heap size until the library is upgraded.

Thanks!

**Comment by Bucimis** [comment]
Thanks for the analysis @alewang,

You are correct - there was a casting-related overflow bug existed until 1.13.5 and was resolved in 1.14.0.  @ChristianToMidiOut are you also running a version of Appboy < 1.14.0?

Although that specific bug should be fixed by upgrading to a version of the SDK > 1.13.5, note that we'll add a further robustness upgrade in our next release which should guard against bad behavior regardless of what `Runtime.getRuntime().maxMemory()` returns.

Cheers
